### PR TITLE
Issue 917

### PR DESCRIPTION
**Pull request by nmorenor**
I think that the code now have addressed your concerns and added a return on promises and now the code is using the concat method of javascript where it fits. I was not aware of this one, a good one to know about, thanks. The changes have been rebased to master so it should be ok to do the pull request now, hopefully this will be enough, let us know your thoughts.

### CODE REVIEW

File: src/git/GitCli.js [hunk 1]
```diff
@@ -545,7 +545,12 @@ define(function (require, exports) {
     function stage(file, updateIndex) {
         var args = ["add"];
         if (updateIndex) { args.push("-u"); }
-        args.push("--", file);
+        if ($.isArray(file)) {
```
Comment (zaggino) [inline src/git/GitCli.js]: this all can be shortened to:
```
function stage(fileOrFiles, updateIndex) {
        var args = ["add"];
        if (updateIndex) { args.push("-u"); }
        return git(args.concat("--", fileOrFiles));
    }
```

File: src/git/GitCli.js [hunk 2]
```diff
@@ -773,6 +778,20 @@ define(function (require, exports) {
         });
     }
 
+    function getDiffOfAllIndexFiles(files) {
+        var args = ["diff", "--no-ext-diff", "--no-color", "--full-index"];
+        if (files !== undefined) {
+            if ($.isArray(files)) {
```
Comment (zaggino) [inline src/git/GitCli.js]: `concat` checks if argument is an `array` and if it's not, it behaves like push so this should be:
```
if (files) {
    args = args.concat("--", files);
}
```

File: src/Utils.js [hunk 3]
```diff
@@ -499,11 +499,13 @@ define(function (require, exports, module) {
...
                             // stage the files again to include stripWhitespace changes
-                            return Git.stage(fileObj.file).then(function () {
+                            if (stageChanges) {
+                                return Git.stage(fileObj.file).then(function () {
 
-                                var t = (new Date()).getTime() - startTime;
-                                notificationDefer.progress(t + "ms - " + Strings.CLEAN_FILE_END + ": " + fileObj.file);
-                            });
+                                    var t = (new Date()).getTime() - startTime;
+                                    notificationDefer.progress(t + "ms - " + Strings.CLEAN_FILE_END + ": " + fileObj.file);
```
Comment (zaggino) [inline src/Utils.js]: notificationDefer should be notified of progress even when we're not staging the changes, no?

File: src/Panel.js [hunk 4]
```diff
@@ -744,9 +803,7 @@ define(function (require, exports) {
                 return Git.resetIndex();
             })
             .then(function () {
-                return Git.stageAll().then(function () {
-                    return handleGitCommit();
-                });
+                return handleGitCommit(undefined, false, COMMIT_MODE.ALL);
```
Comment (zaggino) [inline src/Panel.js]: don't we want to send `lastCommitMessage` here too?

File: src/Panel.js [hunk 5]
```diff
@@ -726,14 +792,7 @@ define(function (require, exports) {
...
             .then(function () {
-                var gitRoot = Preferences.get("currentGitRoot");
-                var currentDoc = DocumentManager.getCurrentDocument();
-                if (currentDoc) {
-                    var relativePath = currentDoc.file.fullPath.substring(gitRoot.length);
-                    return Git.stage(relativePath).then(function () {
-                        return handleGitCommit();
-                    });
-                }
+                return handleGitCommit(undefined, false, COMMIT_MODE.CURRENT);
```
Comment (zaggino) [inline src/Panel.js]: don't we want to send `lastCommitMessage` here too?

**Comment by zaggino** [comment]
Looks good, just a minor stuff, I'll release `0.14.18` now and then merge this for more testing before releasing it into public.

**Comment by nmorenor** [comment]
@zaggino Thanks for the review, we just pushed changes to achieve your comments on the Pull request. Let me know in case you have additional

**Comment by zaggino** [comment]
@nmorenor this is now merged in master, I'll be using the master for a few days (please do the same) before releasing the next version because I want to be 100% sure we didn't break anything.

And thanks for the great work!

**Pull request merged**